### PR TITLE
migrate indexer writer to data ingestion framework

### PR DESCRIPTION
**Pull request by phoenix-o**
notes:
* a new optional method has been added to the `Worker` trait to support the indexer workflow: `preprocess_hook`. It is blocking and enables the execution of certain actions for a task outside of the main `process_checkpoint` routine. It is guaranteed that `preprocess_hook` will be called in order for all checkpoints. In this scenario, it enables the indexing of packages before indexing the checkpoint itself

### CODE REVIEW

File: crates/sui-data-ingestion-core/src/worker_pool.rs [hunk 1]
```diff
@@ -114,6 +114,7 @@ impl<W: Worker + 'static> WorkerPool<W> {
                     if sequence_number < current_checkpoint_number {
                         continue;
                     }
+                    self.worker.preprocess_hook(checkpoint.clone()).expect("failed to preprocess task");
```
Comment (gegaowp) [inline crates/sui-data-ingestion-core/src/worker_pool.rs:117]: are checkpoints from `checkpoint_receiver.recv()` following asc order? iirc download are done in parallel and thus downloaded checkpoints can be out of order, and as a result if a package is published on cp X and then the related objects are updated in X+1, but here it's possible that checkpoint X+1 is assigned a worker before package of cp X is pre-processed, which can be problematic?

**Review by gegaowp — COMMENTED** [review]
I checked, the object indexing via package resolver is indeed the only order sensitive one, left a couple of comments.

File: crates/sui-data-ingestion-core/src/worker_pool.rs (hunk 1 above)
Comment (phoenix-o) [reply crates/sui-data-ingestion-core/src/worker_pool.rs:117]: here all checkpoints are guaranteed to be received in order. Download is done in parallel but results are accumulated before sent to the executor/worker pool

File: crates/sui-indexer/src/handlers/checkpoint_handler.rs [hunk 2]
```diff
@@ -105,121 +106,65 @@ pub struct CheckpointHandler<S> {
     // buffers for packages that are being indexed but not committed to DB,
     // they will be periodically GCed to avoid OOM.
     package_buffer: Arc<Mutex<IndexingPackageBuffer>>,
+    package_resolver: Arc<Resolver<InterimPackageResolver>>,
```
[the remark this answers is not in the document]
Comment (phoenix-o) [reply crates/sui-indexer/src/handlers/checkpoint_handler.rs:109]: let's keep this change for a cleanup PR. I'm not very familiar with the difference and want to minimize number of changes this PR introduces:)

[the remark this answers is not in the document]
Comment (gegaowp) [reply crates/sui-indexer/src/handlers/checkpoint_handler.rs:109]: `package_buffer` is already a field in `package_resolver`, so we can just do `package_resolver.package_buffer`?

File: crates/sui-data-ingestion-core/src/worker_pool.rs (hunk 1 above)
Comment (gegaowp) [reply crates/sui-data-ingestion-core/src/worker_pool.rs:117]: I see, saw that in reader.rs, that's great!

**Review by gegaowp — APPROVED** [review]
Looks good!
regarding rollout, let's benchmark this on a backfill and monitor the backfill speed and memory usage, let's sync offline re the benchmark setup.
please hold on merging this to main though, b/c we have auto release so the change will be picked up and released to production without benchmarking otherwise.

File: crates/sui-indexer/src/indexer.rs [hunk 3]
```diff
@@ -80,16 +81,34 @@ impl Indexer {
...
-            mysten_metrics::metered_channel::ReceiverStream::new(
-                downloaded_checkpoint_data_receiver,
-            ),
-            vec![Box::new(checkpoint_handler)],
-            metrics,
-        )
-        .await;
-
+        #[allow(unused_variables)]
+        let (exit_sender, exit_receiver) = oneshot::channel();
```
Comment (gegaowp) [inline crates/sui-indexer/src/indexer.rs:85]: how is this one-shot channel used?

File: crates/sui-indexer/src/lib.rs [hunk 4]
```diff
@@ -2,16 +2,15 @@
...
-
 use anyhow::{anyhow, Result};
 use clap::Parser;
 use jsonrpsee::http_client::{HeaderMap, HeaderValue, HttpClient, HttpClientBuilder};
 use metrics::IndexerMetrics;
 use mysten_metrics::spawn_monitored_task;
 use prometheus::Registry;
+use std::net::SocketAddr;
 use std::path::PathBuf;
+use std::time::Duration;
```
Comment (gegaowp) [inline crates/sui-indexer/src/lib.rs:13]: I group "use" statements following order of many files in this crate:
std -> other external crates -> sui crates -> this crate

let's keep it this way unless there is some specific reason mergint them?

File: crates/sui-indexer/src/main.rs [hunk 5]
```diff
@@ -19,7 +19,12 @@ async fn main() -> Result<(), IndexerError> {
         .with_env()
         .init();
 
-    let indexer_config = IndexerConfig::parse();
+    let mut indexer_config = IndexerConfig::parse();
+    if indexer_config.rpc_client_url.contains("testnet") {
```
Comment (gegaowp) [inline crates/sui-indexer/src/main.rs:23]: let's add a TODO here to remove the hacky mapping from client url later?

**Review by gegaowp — APPROVED** [review]
LG other than the comments, excited to see this to be complete!

**Pull request merged**